docs/routing.md

```diff
@@ -0,0 +1,248 @@
+
+There are two different types of headers that the build output config can provide. There are important ones, and normal ones. Whether headers should be treated as important is denoted by the presence of an `important` property on a source route. We believe that this indicated that the headers should take precendence over all others on the final response object.
+
+#### Destination
+
+If a source route wants to rewrite the current path to point to a different file, it specifies a destination property. These are taken and the current path is updated, along with applying any search parameters we have collected.
+
+This is where a route's dynamic parameters are applied as search parameters.
+
+When updating the current path to the new destination, additional custom interpretations are needed to ensure compatibility with `RSC` pages. Sometimes, the build output might not generate an `RSC` page, and to avoid breaking routing, we instead want to rewrite the path and return the non-`RSC` page.
+
+In other cases, the build output config may be trying to rewrite a path from `^/` to `/index.rsc`, when the path has additional characters after the `/`. This appears to be a misjudgement in when to apply this rewrite. To ensure that only index pages are affected by this rewrite, we check that the rewritten path is replacing an index path.
+
+#### Checking the New Destination
+
+Sometimes, a route may rewrite a path and require that its existence is verified in the file system before continuing. On the Vercel [documentation](https://vercel.com/docs/build-output-api/v3/configuration#source-route), it states that the `check` property should trigger entering the `filesystem` and `rewrite` phases.
+
+This can be somewhat problematic.
+
+Requests that should be interpreted as requesting the `RSC` version of a page need to enter the `none` phase to receive a rewrite to the `RSC` path. However, if we assume that `check` should take us back to `filesytem`, instead of the routes that aren't in a phase, we would end up not supporting `RSC` pages properly.
+
+Therefore, when the rewritten path has changed, we send the request back to the `none` phase, so that it can be processed from the start.
+
+In the event that the rewritten path is the same as the current path, we opt to send the request to the next phase when the current phase is not `miss`. This custom check for the `miss` phase is necessary in order to prevent infite loops for requests to `/_next/static/...`, and `/_next/data/...`, as they may be rewritten to themselves when calling `check`. As such, if we are in the `miss` phase, we instead update the status code to 404.
+
+#### Continuing with the Routing Process
+
+After all of the above has been completed, the final part to check for a route is whether it would like us to continue matching. If it does not, we have reached the end of matching for this phase, so we can simply exit out with the `done` response.
```

Comment:
```suggestion
After all of the above has been completed, the final part to check for a (matching) route is whether it would like us to continue matching (it does so by setting its `continue` field to `true`). If it does not, we have reached the end of matching for this phase.
```

The `done` response is an implementation details you'd added there right? if so its presence here can make things a bit confusion I think 🤔